### ISSUE TITLE
[2.5.0] Connections unnecessarily closed when returning a no content responses

### ISSUE THREAD

**Issue by studiodev** [playframework/playframework#5821]
### Play Version

Play 2.5 (Play 2.4 is not affected by this bug).
Reproduced with `play.server.netty.transport` = native AND jdk

### API

Scala (not tested with Java)

### Operating System

Tested on :

- Ubuntu 14.04 64bits
- Amazon Linux AMI 

### JDK

java version "1.8.0_74"
Java(TM) SE Runtime Environment (build 1.8.0_74-b02)
Java HotSpot(TM) 64-Bit Server VM (build 25.74-b02, mixed mode)

### Library Dependencies

N/A

### Expected Behavior

Returning a `204 No Content` should not create a connection leak.

### Actual Behavior

When a controller return a 204, the connection seems to not be closed. 
With a `netstat -n`, I can see that the connection is open, stuck in `TIME_WAIT`.

Consequently, running a `wrk -t 32 -c 32 -d10s http://localhost/nocontent` open more than 50000 connections, and the server is not able to respond to any requests until the OS recycle the sockets.
### Reproducible Test Case

Sample project here : https://github.com/studiodev/microservices-benchmark-scala/tree/play25-simplified/microservice-play25

You can easily reproduce the bug by launching a bench on the `/nocontent` route, whereas the `/ok` is not affected.

```
# Terminal 1
watch -n 0,5 'echo NETSTAT> TOTAL: $(netstat -n | grep "127.0.0.1:9000" | wc -l) - TIME_WAIT: $(netstat -n | grep "127.0.0.1:9000" | grep "TIME_WAIT" | wc -l)'

#  Terminal 2

## NOMINAL case
➜  wrk -c32 -t32 -d10s http://localhost:9000/ok
Running 10s test @ http://localhost:9000/ok
  32 threads and 32 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    88.55ms  288.11ms   1.64s    91.77%
    Req/Sec   712.51    413.42     4.55k    66.50%
  192970 requests in 10.10s, 13.80MB read
Requests/sec:  19106.94
Transfer/sec:      1.37MB

NETSTAT> TOTAL: 18 - TIME_WAIT: 18

## ERROR case
➜  wrk -c32 -t32 -d10s http://localhost:9000/nocontent
Running 10s test @ http://localhost:9000/nocontent
  32 threads and 32 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.04ms    4.72ms  79.04ms   90.03%
    Req/Sec   214.03    100.66   515.00     65.50%
  50107 requests in 10.04s, 3.97MB read
  Socket errors: connect 32, read 0, write 0, timeout 0
Requests/sec:   4989.46
Transfer/sec:    404.42KB

NETSTAT> TOTAL: 44196 - TIME_WAIT: 44196
```

I've not tested all the others HTTP code, but it must be the only one (which not expect a body).

**Comment by marcospereira** [comment]
I was not able to reproduce this problem using Debian (`Linux debian 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt11-1 (2015-05-24) x86_64 GNU/Linux`) and the following JDK:
```
java version "1.8.0_72"
Java(TM) SE Runtime Environment (build 1.8.0_72-b15)
Java HotSpot(TM) 64-Bit Server VM (build 25.72-b15, mixed mode)
```
I also tried on Mac OSX (`Darwin Kernel Version 15.3.0`) and can't reproduce it. I even left wkr running for a longer time (5 minutes), but nothing unusual happened.

**Comment by mriehl** [comment]
We are seeing a similar issue, where controller paths that yield a 204 NO CONTENT will get the connection stuck, but only when we POST or PUT on the endpoint with an empty (!) body. I'm not sure if this is helpful to reproduce this issue (TBH we don't have enough data to open an actual issue yet) but there definitely seems to be a problem with connection handling somewhere.

**Comment by studiodev** [comment]
I've updated the sample project with a more minimal use case. I can still reproduce this problem, with just  GET requests.

**Comment by marcospereira** [comment]
Ok. I was (kind of) able to reproduce the problem. I was not able to reach the numbers you did, but definitely there is something wrong:

1. I was able to reproduce the problem on Linux (Linux debian 3.16.0-4-amd64) but not on MacOSX (Darwin Kernel Version 15.3.0)
2. On MacOSX the /nocontent endpoint is slower and has a way smaller throughput. Compare the numbers below.

For all cases I ran `wrk -c82 -t82 -d60s http://localhost:9000/nocontent`:

### `/nocontent`, `/204` and `/304`

Running `wrk -c82 -t82 -d60s http://localhost:9000/nocontent`:
```
Running 1m test @ http://localhost:9000/nocontent
  82 threads and 82 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.08ms    8.68ms 350.10ms   99.86%
    Req/Sec    87.81     27.99   121.00     86.44%
  32921 requests in 1.00m, 2.61MB read
  Socket errors: connect 0, read 82, write 0, timeout 0
Requests/sec:    547.87
Transfer/sec:     44.41KB
```

### `/ok`

```
Running 1m test @ http://localhost:9000/ok
  82 threads and 82 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.06ms    6.75ms 188.63ms   98.29%
    Req/Sec   740.14    250.65     6.53k    91.20%
  3233002 requests in 1.00m, 231.24MB read
  Socket errors: connect 0, read 0, write 0, timeout 36
Requests/sec:  53793.29
Transfer/sec:      3.85MB
```

### `/200`

```
Running 1m test @ http://localhost:9000/200
  82 threads and 82 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.87ms    5.23ms 152.39ms   98.74%
    Req/Sec   698.35     92.30     4.20k    90.44%
  3421503 requests in 1.00m, 244.72MB read
Requests/sec:  56967.82
Transfer/sec:      4.07MB
```

So, yes. There is a problem with responses that have no body.

**Comment by jroper** [comment]
I've updated the title to reflect the true problem here.  There is no connection leak - `TIME_WAIT` connections are connections that have been closed properly, however TCP requires that the closing side of a connection keep a knowledge of the connection for a certain amount of time (platform dependent, eg 60 seconds) after closing it, in case the final `ACK` didn't get through and the remote end resends a `FIN ACK`.  The state that the socket is in when its in this state is called `TIME_WAIT`, and it will go out of this state after the timeout, this is normal operation, so there's no connection leak.

What the problem is is that when the body has no content length or transfer encoding, Play is disabling keep alive - ie, setting the `Connection: close` header, and then closing the connection once the empty body is sent.  For most response statuses, this is necessary, since when no content length or transfer encoding is prescribed, the only way to signal that the response body is ended is to close the connection.  However, for response statuses that can't have a response body, these don't need (and shouldn't have) a content length or transfer encoding, but since they can't have a response body, you don't need to close the connection.  So, Play is unnecessarily disabling keep alive, and closing the connection, for response bodies with no content.

Since the connection is closed, and must be reestablished between every request, this explains the dismal performance.  And of course, all these closed connections stay in `TIME_WAIT` for 60 seconds, limiting the number of requests of this type that a single host can make to an upper bound of 65535 req/min.
